drivers/clock_control/clock_stm32_mco.c

```diff
@@ -0,0 +1,136 @@
+#elif DT_PROP(DT_NODELABEL(mco1), source) == STM32_SRC_PLLQCLK
+	#if (CONFIG_SOC_SERIES_STM32G0X || CONFIG_SOC_SERIES_STM32WLX)
+		#define MCO1_SOURCE	LL_RCC_MCO1SOURCE_PLLQCLK
+	#elif (CONFIG_SOC_SERIES_STM32H5X || \
+		 CONFIG_SOC_SERIES_STM32H7X || CONFIG_SOC_SERIES_STM32H7RSX)
+		#define MCO1_SOURCE	LL_RCC_MCO1SOURCE_PLL1QCLK
+	#else
+		#error "PLLQCLK is not a valid clock source on your SOC"
+	#endif
+#elif DT_PROP(DT_NODELABEL(mco1), source) == STM32_SRC_PLLCLK_DIV2
+	#define MCO1_SOURCE		LL_RCC_MCO1SOURCE_PLLCLK_DIV_2
+#elif DT_PROP(DT_NODELABEL(mco1), source) == STM32_SRC_PLL2CLK
+	#define MCO1_SOURCE		LL_RCC_MCO1SOURCE_PLL2CLK
+#elif DT_PROP(DT_NODELABEL(mco1), source) == STM32_SRC_PLLI2SCLK
+	#define MCO1_SOURCE		LL_RCC_MCO1SOURCE_PLLI2SCLK
+#elif DT_PROP(DT_NODELABEL(mco1), source) == STM32_SRC_PLLI2SCLK_DIV2
+	#define MCO1_SOURCE		LL_RCC_MCO1SOURCE_PLLI2SCLK_DIV2
+#elif DT_PROP(DT_NODELABEL(mco1), source) == STM32_SRC_SYSCLK
+	#define MCO1_SOURCE		LL_RCC_MCO1SOURCE_SYSCLK
+#else
+#error "MCO1 source not a valid clock source"
+#endif
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(mco1), okay) */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(mco2), okay)
+#if DT_PROP(DT_NODELABEL(mco2), source) == NO_SEL
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_SYSCLK
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_PLLI2S
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_PLLI2S
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_HSE
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_HSE
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_LSI
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_LSI
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_CSI
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_CSI
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_PLLCLK
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_PLLCLK
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_PLLPCLK
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_PLL1PCLK
+#elif DT_PROP(DT_NODELABEL(mco2), source) == STM32_SRC_PLL2PCLK
+	#define MCO2_SOURCE		LL_RCC_MCO2SOURCE_PLL2PCLK
+#else
+#error "MCO2 source not a valid clock source"
+#endif
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(mco2), okay) */
+
+#define fn_mco_prescaler(i, v) LL_RCC_MCO ## i ## _DIV_ ## v
+#define mco_prescaler(i, v) fn_mco_prescaler(i, v)
+
+#define fn_mco_source(i) MCO ## i ## _SOURCE
+#define mco_source(i) fn_mco_source(i)
+
+struct stm32_mco_config {
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct stm32_mco_data {
+	uint32_t source;
+	uint32_t prescaler;
+};
+
+static int stm32_mco_init(const struct device *dev)
+{
+	const struct stm32_mco_config *config = dev->config;
+	struct stm32_mco_data *data = dev->data;
+
+	(void)pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+	LL_RCC_ConfigMCO(data->source);
+#else
+	LL_RCC_ConfigMCO(data->source, data->prescaler);
+#endif
+
+	return 0;
+}
+
+#define STM32_MCO_INIT(inst)                                            \
+									\
```

Comment:
It appears we lose the prescaler restriction provided by Kconfig `range`. Could you re-integrate it here?

While we _could_ add manual `#if SERIES_xxx { BUILD_ASSERT(1 <= prescaler <= XX }` checks, I think we might be able to make a generic check like:
```c
//place this before the STM32_MCO_INIT macro definition

/* Use a macro to have everything expanded in the error message
 * IS_ENABLED() here serves as a moral equivalent to C preprocessor
 * `defined()`, which is only available in #if checks
 */
#define ASSERT_VALID_PRESCALER(inst) \
        BUILD_ASSERT(IS_ENABLED(mco_prescaler(inst)),
                     "Invalid prescaler selected for MCO instance " #inst
                     " (" DT_NODE_FULL_NAME(DT_DRV_INST(inst))  ")");
```

```suggestion
									\
ASSERT_VALID_PRESCALER(inst);                                           \
```

Reply:
Just to be clear, giving an invalid prescaler value would end in compilation error, because the definition does not exists, which is the same reason why your suggested IS_ENABLED also work.

This suggestion does give an explanation, so it is better.
The resulting macro expansion explanation is still horrible though.

Reply:
I cannot get this suggestion to work. I will add another check, that is not as elegant, but does give a proper explanation to the user.

Reply:
Good catch. Kconfig is limited to merely `range` - since it's better than nothing, I guess it was added, but did not map perfectly to all series' hardware.

My suggestion was indeed to provide a clearer error message, since, as you said, invalid prescalers will break the build regardless as the LL macro will be missing.

Since my suggestion doesn't work, ignore this comment as this is mostly cosmetic.

Reply:
With the new approach, the user needs to give the divisor value for the register directly.
For each platform we can make defines for the supported values.
There is a lot of variance in what is supported or not.

Reply:
Users need to go and look up the values in the reference manual though, as you can see how this looks in the sample.
The choices and values are different for every device. But can be added later, I will not do it now.